### PR TITLE
feat: add `Reducible.{min, max, dropWhile, takeWhile, intersperse, toXYZ}`

### LINKED ISSUE #1678: Type Classes: Add Reducible Type Class
`Reducible` is like `Foldable` on a non-empty collection.

Propose a design.

See http://typelevel.org/cats/api/cats/Reducible.html

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes #1678

### CODE REVIEW

**Comment by jaschdoc** [comment]
RFR. I've made some questionable changes to `List` in my attempt to override some instances... not sure what you think and we can easily drop those overrides + the changes.

File: main/src/library/List.flix [hunk 1]
```diff
@@ -223,33 +223,47 @@ namespace List {
...
     }
 
     ///
     /// Optionally returns the first element of `l` that satisfies the predicate `f` when searching from right to left.
     ///
-    /// The function `f` must be pure.
-    ///
     @Time(time(f) * length(l)) @Space(space(f))
-    pub def findRight(f: a -> Bool, l: List[a]): Option[a] = match l {
+    pub def findRight(f: a -> Bool & ef, l: List[a]): Option[a] & ef =
```
Comment (magnus-madsen) [inline main/src/library/List.flix:242]: Whats the purpose of this change? Was the old code not fine (other than not being in CPS?)

File: main/src/library/Nel.flix [hunk 2]
```diff
@@ -79,10 +79,24 @@ instance Iterable[Nel] {
...
 }
-
 instance Reducible[Nel] {
     pub def reduceLeftTo(f: (b, a) -> b & ef1, g: a -> b & ef2, l: Nel[a]): b & (ef1 and ef2) = Nel.reduceLeftTo(f, g, l)
     pub def reduceRightTo(f: (a, b) -> b & ef1, g: a -> b & ef2, l: Nel[a]): b & (ef1 and ef2) = Nel.reduceRightTo(f, g, l)
+    override pub def head(l: Nel[a]): a = Nel.head(l)
+    override pub def last(l: Nel[a]): a = Nel.last(l)
+    override pub def init(l: Nel[a]): List[a] = Nel.init(l)
+    override pub def tail(l: Nel[a]): List[a] = Nel.tail(l)
+    override pub def exists(f: a -> Bool & ef, l: Nel[a]): Bool & ef = Nel.exists(f, l)
```
Comment (magnus-madsen) [inline main/src/library/Nel.flix:89]: I understand overriding the above signatures, but why override e.g. memberOf?

Comment (jaschdoc) [reply main/src/library/Nel.flix:89]: Same as `exists` and `forall`; to allow early termination

File: main/src/library/List.flix (hunk 1 above)
Comment (jaschdoc) [reply main/src/library/List.flix:242]: `f` was not effect polymorphic. Had to be pure

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**